lib/SparkPost/APIResource.php

```diff
@@ -1,216 +1,214 @@
+   * @param array $options (optional) query string parameters
+   * @return array Result of the request
+   */
+  public function get( $resourcePath=null, Array $query=[] ) {
+    return $this->callResource( 'get', $resourcePath, ['query'=>$query] );
+  }
+
+  /**
+   * @desc Wrapper method for issuing DELETE request to current API endpoint
+   *
+   * @param string $resourcePath (optional) string resource path of specific resource
+   * @param array $options (optional) query string parameters
+   * @return array Result of the request
+   */
+  public function delete( $resourcePath=null, Array $query=[] ) {
+    return $this->callResource( 'delete', $resourcePath, ['query'=>$query] );
+  }
+
+
+  /**
+   * @desc assembles a URL for a request
+   * @param string $resourcePath path after the initial endpoint
+   * @param array options array with an optional value of query with values to build a querystring from.
+   * @return string the assembled URL
+   */
+  private function buildUrl($resourcePath, $options) {
+    $url = join(['/', $this->endpoint, '/']);
+    if (!is_null($resourcePath)){
+      $url .= $resourcePath;
+    }
+
+    if( !empty($options['query'])) {
+      $queryString = http_build_query($options['query']);
+      $url .= '?'.$queryString;
+    }
+
+    return $url;
+  }
+
+
+  /**
+   * @desc Prepares a body for put and post requests
+   * @param array options array with an optional value of body with values to build a request body from.
+   * @return string|null A json encoded string or null if no body was provided
+   */
+  private function buildBody($options) {
+    $body = null;
+    if( !empty($options['body']) ) {
+      $model = static::$structure;
+      $requestModel = $this->buildRequestModel( $options['body'], $model );
+      $body = json_encode($requestModel);
+    }
+    return $body;
+  }
+
+
+  /**
+   * @desc Private Method for issuing GET and DELETE request to current API endpoint
+   *
+   *  This method is responsible for getting the collection _and_
+   *  a specific entity from the API endpoint
+   *
+   *  If resourcePath parameter is omitted, then we fetch the collection
+   *
+   * @param string $action HTTP method type
+   * @param string $resourcePath (optional) string resource path of specific resource
+   * @param array $options (optional) query string parameters
+   * @return array Result set of action performed on resource
+   */
+  private function callResource( $action, $resourcePath=null, $options=[] ) {
+    $action = strtoupper($action); // normalize
+
+    if( !in_array($action, ['POST', 'PUT', 'GET', 'DELETE'])) {
+      throw new \Exception('Invalid resource action');
+    }
+
+    $url = $this->buildUrl($resourcePath, $options);
+    $body = $this->buildBody($options);
+
+    //make request
+    try {
+      $response = $this->sparkpost->httpAdapter->send($url, $action, $this->sparkpost->getHttpHeaders(), $body);
```

Comment:
I guess yeah you could allow $headers to be passed in via $options or something else ... probably not necessary right now but this method is already a little noisy so no harm in leaving the getHttpHeaders as a separate function if you want to, I don't really have an opinion either way /shrug